### PR TITLE
upgrade testing for concurrent preburns

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds a number of tests to test the upgrade path for concurrent preburns in the new Diem Framework. 

In order to test this, some new testing infrastructure was added to support starting the `FakeExecutor` with the test genesis at release 1.1. Note that we need to also save the privkey/pubkey pair for the release genesis as we can't reuse the new ones that we generate, since the underlying `StdRng` implementation changed in the updated `rand` crate. 

### Code updates
This updates the upgrade flow for preburns to allow DDs to upgrade to a preburn queue even if they have a non-empty preburn at the time of the call that upgrade the `Preburn` resource.

### Spec updates
The spec on `upgrade_preburn` has been updated to handle the new logic. 

### Tests
There are a number of upgrade paths checked in the new tests, these are:
* Upgrade with an empty preburn
* Upgrade with a non-empty preburn
* Old script on old DF and old DF data
* Old scripts on new DF and old DF data
* New scripts on old DF and old DF data
* New scripts on new DF and old DF data
* New scripts on new DF and new DF data

### CODE REVIEW

File: language/diem-framework/modules/Diem.move [hunk 1]
```diff
@@ -679,7 +684,27 @@ module Diem {
         let account_addr = Signer::spec_address_of(account);
         let preburn_exists = exists<Preburn<CoinType>>(account_addr);
         let preburn_queue_exists = exists<PreburnQueue<CoinType>>(account_addr);
-        include (preburn_exists && !preburn_queue_exists) ==> PublishPreburnQueueEnsures<CoinType>;
+        let preburn = global<Preburn<CoinType>>(account_addr);
```
Comment (tzakian) [inline language/diem-framework/modules/Diem.move:687]: @wrwg @shazqadeer let me know if these changes look OK for this PR.

File: language/diem-framework/modules/Diem.move [hunk 2]
```diff
@@ -651,8 +651,14 @@ module Diem {
...
             let Preburn { to_burn } = move_from<Preburn<CoinType>>(sender);
-            destroy_zero(to_burn);
             publish_preburn_queue<CoinType>(account);
+            // If the DD has an old preburn balance, this is converted over
+            // into the new preburn queue when it's upgraded.
+            if (to_burn.value > 0)  {
+                add_preburn_to_queue(account, Preburn { to_burn })
+            } else {
+                destroy_zero(to_burn)
+            };
```
Comment (tzakian) [inline language/diem-framework/modules/Diem.move:661]: @moeziniaf Let me know if these changes look good to you.

This now allows a DD to upgrade even if they have a full preburn from before the Diem Framework was updated (I'll cc you in the test case that demonstrates this).

File: language/e2e-testsuite/src/tests/preburn_queue.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+        burn_with_amount_new(&mut executor, 100, &tc_account, &mut tc_seqno, true)
+            .unwrap()
+            .status()
+            .status()
+            .unwrap(),
+        KeptVMStatus::MoveAbort(_, 2821)
+    ));
+
+    // Now upgrade the preburn (recall it already has a nonzero balance).
+    create_preburn_balance(&mut executor, &dd_account, 100, &mut dd_seqno, false);
```
Comment (tzakian) [inline language/e2e-testsuite/src/tests/preburn_queue.rs:296]: @moeziniaf this is the case that the new code is handling - an existing preburn request at the time of the DF upgrade.

**Comment by tzakian** [comment]
## Main reviewers: 

@runtian-zhou for testing and release

@moeziniaf for AOS and behavior changes in the upgrade

@wrwg and @shazqadeer for spec changes in `upgrade_preburn`

File: language/diem-framework/modules/Diem.move (hunk 2 above)
Comment (moeziniaf) [reply language/diem-framework/modules/Diem.move:661]: great - this should give us the guarantee there are either 0 or 1 preburns in the Queue after the upgrade.

Comment (moeziniaf) [reply language/diem-framework/modules/Diem.move:661]: (If we wanted to be extra pedantic, we could make this check for the upgrade case too but again that piece of code `add_preburn_to_queue_first_time` would be a one time piece of code)

File: language/e2e-testsuite/src/tests/preburn_queue.rs (hunk 3 above)
Comment (moeziniaf) [reply language/e2e-testsuite/src/tests/preburn_queue.rs:295]: The code dealing with the df_upgrade, could it be used for testing other upgrade paths? Not sure exactly how all the new test executor stuff works but might we want to document it and put it in a common place for other e2e tests which may need it?

**Review by moeziniaf — APPROVED** [review]

File: language/e2e-testsuite/src/tests/preburn_queue.rs [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+        let compiled_module_id = CompiledModule::deserialize(&compiled_module_bytes)
+            .unwrap()
+            .self_id();
+        executor.add_module(&compiled_module_id, compiled_module_bytes);
+    }
+
+    executor.execute_and_apply(
+        dr_account
+            .transaction()
+            .script(encode_update_diem_version_script(0, 2))
```
Comment (moeziniaf) [inline language/e2e-testsuite/src/tests/preburn_queue.rs:72]: well this is handy :)

File: language/e2e-testsuite/src/tests/preburn_queue.rs (hunk 3 above)
Comment (tzakian) [reply language/e2e-testsuite/src/tests/preburn_queue.rs:229]: Good point! I've updated the PR pulling this functionality out :)

**Review by wrwg — APPROVED** [review]

**Pull request closed without merging**